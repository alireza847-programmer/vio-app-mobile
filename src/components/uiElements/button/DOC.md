# VButton Component Documentation

## Overview

The `VButton` component is a versatile button element designed for React Native applications. It provides flexibility in styling and configuration to suit different design needs.

## Usage

```jsx
import React from 'react';
import VButton, {VButtonProps} from './path-to-your-component/VButton'; // Adjust the import path based on your project structure

const MyComponent = () => {
  const handlePress = () => {
    // Handle button press logic here
  };

  const buttonProps: VButtonProps = {
    mode: 'primary', // Options: 'primary', 'secondary', 'error'
    styled: 'FULL_WIDTH', // Options: 'TEXT', 'FULL_WIDTH', 'CONTENT_SIZE'
    title: 'Click me',
    subTitle: 'Optional subtitle',
    onPress: handlePress,
    // Add other props as needed
  };

  return <VButton {...buttonProps} />;
};
```

## Props

### VButton Props

- `mode` (string): The mode of the button. Options include 'primary', 'secondary', and 'error'.
- `styled` (string): The style of the button. Options include 'TEXT', 'FULL_WIDTH', and 'CONTENT_SIZE'.
- `width` (DimensionValue): The width of the button. Use this to set a custom width for the button.

### Extended VButton Props

- `title` (string): The main text content of the button.
- `icon` (function): A function that returns a React element to be used as the button icon.
- `rightIcon` (function): A function that returns a React element to be used as the icon on the right side of the button.
- `subTitle` (string): Optional subtitle text to display next to the main title.
- `onPress` (function): A callback function triggered when the button is pressed.
- `testID` (string): Test identifier for testing purposes.
- `disabled` (boolean): If true, the button is in a disabled state.

## Example

```jsx
import React from 'react';
import VButton, {VButtonProps} from './path-to-your-component/VButton'; // Adjust the import path based on your project structure

const ExampleScreen = () => {
  const handlePress = () => {
    // Handle button press logic here
  };

  const buttonProps: VButtonProps = {
    mode: 'primary',
    title: 'Click me',
    subTitle: 'Optional subtitle',
    styled: 'FULL_WIDTH',
    onPress: handlePress,
  };

  return <VButton {...buttonProps} />;
};

export default ExampleScreen;
```

## Icons

The `icon` and `rightIcon` props allow you to add custom icons to the left and right sides of the button, enhancing its visual appeal. Pass functions that return React elements for customization.
