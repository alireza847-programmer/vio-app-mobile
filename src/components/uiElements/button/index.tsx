import React from 'react';
import Button from './style';
import {VButtonProps} from 'types/components/uiElements/button';
import VText from '../text';
import {theme} from 'themes/emotion';
import VRow from '../row';

const VButton = (props: VButtonProps) => {
  const {
    mode = 'primary',
    subTitle,
    title,
    styled = 'FULL_WIDTH',
    icon,
    rightIcon,
    onPress,
    testID,
    disabled = false,
    width,
  } = props;

  const renderContentColor = () => {
    if (disabled) {
      return theme.colors.textColor.button.disabled;
    }
    if (styled === 'TEXT') {
      return theme.colors.button[mode];
    }
    return theme.colors.textColor.button[mode];
  };

  return (
    <Button
      width={width}
      disabled={disabled}
      testID={testID}
      activeOpacity={0.8}
      onPress={onPress}
      styled={styled}
      mode={mode}>
      {icon && (
        <VRow {...(title && {marginRighRatio: 2})} fullWidth={false}>
          {icon(renderContentColor())}
        </VRow>
      )}
      <VText
        typography="semiBold16"
        color={renderContentColor()}
        marginTopRatio={0}>
        {title}
      </VText>
      {subTitle && (
        <VText
          typography="regular16"
          color={renderContentColor()}
          marginLeftRatio={1.2}
          marginTopRatio={0}>
          {subTitle}
        </VText>
      )}
      {rightIcon && (
        <VRow {...(title && {marginLeftRatio: 2})} fullWidth={false}>
          {rightIcon(renderContentColor())}
        </VRow>
      )}
    </Button>
  );
};
export default VButton;
