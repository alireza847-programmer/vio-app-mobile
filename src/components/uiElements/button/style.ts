import styled from '@emotion/native';
import {ButtonProps} from 'types/components/uiElements/button';

const Button = styled.TouchableOpacity<ButtonProps>(
  ({theme, mode = 'primary', styled = 'FULL_WIDTH'}) => ({
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    borderRadius: theme.radius * 2,
    borderColor: theme.colors.stroke.button,
    ...(styled === 'FULL_WIDTH' && {width: '100%'}),
    ...(styled === 'CONTENT_SIZE' && {
      paddingHorizontal: theme.HorizontalRem * 4,
    }),
    ...(styled !== 'TEXT' && {
      paddingVertical: theme.verticalRem * 4,
    }),
    ...(styled !== 'TEXT' && {backgroundColor: theme.colors.button[mode]}),
    ...(mode === 'secondary' && styled !== 'TEXT' && {borderWidth: 1}),
  }),
);

export default Button;
