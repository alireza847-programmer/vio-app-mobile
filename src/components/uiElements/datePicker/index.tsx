import {theme} from '@storybook/react-native';
import React from 'react';
import {View} from 'react-native';
import Calendar from 'react-native-calendar-range-picker';
import {VDatePickerProps} from 'types/components/uiElements/datePicker';

const VDatePicker = (props: VDatePickerProps) => {
  const {...rest} = props;
  return (
    <View style={{width: '100%'}}>
      <Calendar disabledBeforeToday pastYearRange={0} {...rest} />
    </View>
  );
};

export default VDatePicker;
