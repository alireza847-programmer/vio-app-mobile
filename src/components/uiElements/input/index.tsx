import React from 'react';
import Input from './style';
import {VInputProps} from 'types/components/uiElements/input';
import {theme} from 'themes/emotion';

const VInput = (props: VInputProps) => {
  const {...rest} = props;
  return (
    <Input
      placeholderTextColor={theme.colors.textColor.placeHolder}
      {...rest}
    />
  );
};

export default VInput;
