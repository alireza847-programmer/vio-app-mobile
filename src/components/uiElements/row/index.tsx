import styled from '@emotion/native';
import {PropsWithChildren} from 'react';
import {RowProps} from 'types/components/uiElements/row';

const VRow = styled.View<PropsWithChildren<RowProps>>(
  ({
    theme,
    backgroundColor,
    fullWidth = true,
    justifyContent = 'flex-start',
    alignItems = 'center',
    marginTopRatio = 0,
    marginLeftRatio = 0,
    paddingHorizontalRatio = 0,
    paddingVerticalRatio = 0,
    marginRighRatio = 0,
    minWidth,
  }) => ({
    flexDirection: 'row',
    justifyContent,
    alignItems,
    marginTop: theme.verticalRem * marginTopRatio,
    marginLeft: theme.HorizontalRem * marginLeftRatio,
    marginRight: theme.HorizontalRem * marginRighRatio,
    paddingHorizontal: theme.HorizontalRem * paddingHorizontalRatio,
    paddingVertical: theme.verticalRem * paddingVerticalRatio,
    ...(backgroundColor && {backgroundColor}),
    ...(fullWidth && {width: '100%'}),
    ...(minWidth && {minWidth}),
  }),
);

export default VRow;
