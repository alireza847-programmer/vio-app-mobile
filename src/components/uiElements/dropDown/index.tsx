import React, {Fragment, useEffect, useState} from 'react';
import {VDropDownProps} from 'types/components/uiElements/dropDown';
import {theme} from 'themes/emotion';
import {ChevronSvg} from 'assets/svgs';
import {Container} from './style';
import WheelPicker from 'react-native-wheely';
import VModal from '../modal';
import VText from '../text';
import VRow from '../row';
import VButton from '../button';
import {TouchableOpacity} from 'react-native';

const VDropDown = (props: VDropDownProps) => {
  const {data, placeHolder, onChange, defaultValue} = props;
  const [visibleModal, setVisibleModal] = useState(false);
  const [value, setValue] = useState('');
  const [selectedIndex, setSelectedIndex] = useState(0);
  useEffect(() => {
    if (defaultValue) {
      setValue(String(defaultValue));
      setSelectedIndex(data.indexOf(String(defaultValue)));
    } else {
      setValue(String(placeHolder));
    }
  }, [placeHolder, defaultValue]);
  const onConfirm = () => {
    setVisibleModal(false);
    setValue(data[selectedIndex]);
    onChange(data[selectedIndex]);
  };
  const renderModal = () => {
    return (
      <VModal
        withClose
        withHeader
        title="Child 1 age"
        onClose={() => setVisibleModal(false)}
        isVisible={visibleModal}>
        <WheelPicker
          selectedIndex={selectedIndex}
          options={data}
          onChange={index => setSelectedIndex(index)}
          containerStyle={{width: '90%'}}
          opacityFunction={x => 0.3}
          selectedIndicatorStyle={{
            backgroundColor: theme.colors.button.secondary,
            borderTopWidth: 1,
            borderBottomWidth: 1,
            borderColor: theme.colors.button.primary,
          }}
          visibleRest={1}
          itemStyle={{
            backgroundColor: 'transparent',
          }}
          itemTextStyle={{
            ...theme.fonts.semiBold18,
          }}
        />
        <VRow paddingVerticalRatio={4} justifyContent="center">
          <VButton onPress={onConfirm} width={'90%'} title="CONFIRM" />
        </VRow>
      </VModal>
    );
  };
  return (
    <Fragment>
      <TouchableOpacity
        activeOpacity={0.8}
        onPress={() => setVisibleModal(true)}>
        <Container
          alignItems="center"
          justifyContent="space-between"
          fullWidth={false}>
          <VText
            width={theme.HorizontalRem * 10}
            textAlign="left"
            typography="semiBold16">
            {value}
          </VText>
          <VRow fullWidth={false} marginLeftRatio={3}>
            <ChevronSvg fill={theme.colors.button.primary} />
          </VRow>
        </Container>
      </TouchableOpacity>
      {renderModal()}
    </Fragment>
  );
};
export default VDropDown;
