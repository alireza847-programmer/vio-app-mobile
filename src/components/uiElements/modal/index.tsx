import Modal from 'react-native-modal';
import React, {PropsWithChildren} from 'react';
import {VModalProps} from 'types/components/uiElements/modal';
import {HeaderContainer, Wrapper, modalStyle} from './style';
import {theme} from 'themes/emotion';
import {CloseSvg} from 'assets/svgs';
import VText from '../text';
import VButton from '../button';

const VModal = (props: PropsWithChildren<VModalProps>) => {
  const {children, isVisible, onClose, title, withClose, withHeader} = props;
  return (
    <Modal
      animationIn="slideInUp"
      animationOut={'slideOutDown'}
      style={modalStyle}
      testID="modal"
      hasBackdrop={true}
      onBackButtonPress={onClose}
      onBackdropPress={() => onClose()}
      isVisible={isVisible}>
      <Wrapper>
        {withHeader && (
          <HeaderContainer
            justifyContent="space-between"
            paddingHorizontalRatio={4}
            paddingVerticalRatio={6}>
            {withClose && (
              <VButton
                onPress={onClose}
                testID="close-button"
                styled="TEXT"
                icon={fill => <CloseSvg fill={theme.colors.button.primary} />}
              />
            )}
            <VText typography="semiBold16">{title}</VText>
            <CloseSvg fill={'transparent'} />
          </HeaderContainer>
        )}
        {children}
      </Wrapper>
    </Modal>
  );
};
export default VModal;
