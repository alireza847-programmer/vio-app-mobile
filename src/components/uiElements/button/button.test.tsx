import React from 'react';
import {render} from '@testing-library/react-native';
import VButton from './index';
import {it, describe, expect} from '@jest/globals';
import {VButtonProps} from 'types/components/uiElements/button';
import {ThemeProvider} from '@emotion/react';
import {theme} from 'themes/emotion';

// A wrapper component to provide the theme context to VButton
const Component = (props: VButtonProps) => (
  <ThemeProvider theme={theme}>
    <VButton {...props} />
  </ThemeProvider>
);

describe('VButton component', () => {
  it('renders correctly with default props', () => {
    // Arrange
    const {toJSON} = render(<Component onPress={() => {}} />);

    // Assert
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly with custom title and mode', () => {
    // Arrange
    const {toJSON} = render(
      <Component onPress={() => {}} title="Click me" mode="secondary" />,
    );

    // Assert
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly with custom styled prop', () => {
    // Arrange
    const {toJSON} = render(
      <Component title="Click Me" onPress={() => {}} styled="CONTENT_SIZE" />,
    );

    // Assert
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly with custom icon and rightIcon', () => {
    // Arrange
    const MockIcon = () => <svg>Mock Icon</svg>;
    const {toJSON} = render(
      <Component
        onPress={() => {}}
        icon={() => <MockIcon />}
        rightIcon={() => <MockIcon />}
      />,
    );

    // Assert
    expect(toJSON()).toMatchSnapshot();
  });
});
