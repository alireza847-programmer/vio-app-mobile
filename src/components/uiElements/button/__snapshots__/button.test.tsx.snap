// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VButton component matches snapshot with custom icon and rightIcon 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#0071F3",
      "borderColor": "#DAE9FA",
      "borderRadius": 15,
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "paddingVertical": 26.68,
      "width": "100%",
    }
  }
>
  <View
    fullWidth={false}
    style={
      [
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "flex-start",
          "marginLeft": 0,
          "marginRight": 0,
          "marginTop": 0,
          "paddingHorizontal": 0,
          "paddingVertical": 0,
        },
        undefined,
      ]
    }
  >
    <svg>
      Mock Icon
    </svg>
  </View>
  <Text
    color="#FFFFFF"
    marginTopRatio={0}
    style={
      [
        {
          "color": "#FFFFFF",
          "fontFamily": "Inter-SemiBold",
          "fontSize": 16,
          "marginLeft": 0,
          "marginTop": 0,
          "textAlign": "center",
          "textAlignVertical": "center",
        },
        undefined,
      ]
    }
    typography="semiBold16"
  />
  <View
    fullWidth={false}
    style={
      [
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "flex-start",
          "marginLeft": 0,
          "marginRight": 0,
          "marginTop": 0,
          "paddingHorizontal": 0,
          "paddingVertical": 0,
        },
        undefined,
      ]
    }
  >
    <svg>
      Mock Icon
    </svg>
  </View>
</View>
`;

exports[`VButton component matches snapshot with custom styled prop 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#0071F3",
      "borderColor": "#DAE9FA",
      "borderRadius": 15,
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "paddingHorizontal": 30,
      "paddingVertical": 26.68,
    }
  }
>
  <Text
    color="#FFFFFF"
    marginTopRatio={0}
    style={
      [
        {
          "color": "#FFFFFF",
          "fontFamily": "Inter-SemiBold",
          "fontSize": 16,
          "marginLeft": 0,
          "marginTop": 0,
          "textAlign": "center",
          "textAlignVertical": "center",
        },
        undefined,
      ]
    }
    typography="semiBold16"
  >
    Click Me
  </Text>
</View>
`;

exports[`VButton component matches snapshot with custom title and mode 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#F7FBFF",
      "borderColor": "#DAE9FA",
      "borderRadius": 15,
      "borderWidth": 1,
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "paddingVertical": 26.68,
      "width": "100%",
    }
  }
>
  <Text
    color="#0071F3"
    marginTopRatio={0}
    style={
      [
        {
          "color": "#0071F3",
          "fontFamily": "Inter-SemiBold",
          "fontSize": 16,
          "marginLeft": 0,
          "marginTop": 0,
          "textAlign": "center",
          "textAlignVertical": "center",
        },
        undefined,
      ]
    }
    typography="semiBold16"
  >
    Click me
  </Text>
</View>
`;

exports[`VButton component matches snapshot with default props 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#0071F3",
      "borderColor": "#DAE9FA",
      "borderRadius": 15,
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "paddingVertical": 26.68,
      "width": "100%",
    }
  }
>
  <Text
    color="#FFFFFF"
    marginTopRatio={0}
    style={
      [
        {
          "color": "#FFFFFF",
          "fontFamily": "Inter-SemiBold",
          "fontSize": 16,
          "marginLeft": 0,
          "marginTop": 0,
          "textAlign": "center",
          "textAlignVertical": "center",
        },
        undefined,
      ]
    }
    typography="semiBold16"
  />
</View>
`;
